File: real-world-functional-programming/Chapter16/Chapter16_LinqEvents/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Chapter16_LinqEvents
{
  public partial class MainForm : Form
  {
    public MainForm()
    {
      InitializeComponent();
    }

    private void MainForm_Load(object sender, EventArgs e)
    {
      // Convert .NET events to Reactive LINQ event values
      var upEvent = Observable.FromEvent<EventArgs>(btnUp, "Click");
      var downEvent = Observable.FromEvent<EventArgs>(btnDown, "Click");

      // Create events carrying -1 or +1 as values
      var up = from clickArgs in upEvent select +1;
      var down = from clickArgs in downEvent select -1;

      // Merge events 
      Observable.Merge(up, down)
        // Calculate sum of values carried by the event
        .Scan(0, (state, num) => state + num)
        .Subscribe(sum =>
          lblCount.Text = string.Format("Count: {0}", sum));
    }
  }
}

File: real-world-functional-programming/Chapter16/Chapter16_VbEvents/MainForm.vb
Imports System.Linq
Imports Chapter16_VbEvents.ObservableExtensions

Public Class MainForm

  Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    ' Convert .NET events to Reactive LINQ event values
    Dim upEvent = Observable.FromEvent(Of EventArgs)(btnUp, "Click")
    Dim downEvent = Observable.FromEvent(Of EventArgs)(btnDown, "Click")

    ' Create events carrying +1/-1, merge them and aggregate them using 'Sum'
    Dim sumEvent = _
      Aggregate num In Observable.Merge( _
        (From clickArgs1 In upEvent Select +1), _
        (From clickArgs2 In downEvent Select -1)) _
      Into MovingSum(num)
    sumEvent.Subscribe(AddressOf UpdateCount)

  End Sub

  Sub UpdateCount(ByVal sum As Integer)
    lblCount.Text = String.Format("Count: {0}", sum)
  End Sub

End Class
